flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/over/NonTimeUnboundedPrecedingFunction.java

```diff
@@ -0,0 +1,482 @@
+        } else {
+            prevAcc = accMapState.get(ids.get(ids.size() - 2));
+        }
+
+        while (i < ids.size()) {
+            RowData curValue = valueMapState.get(ids.get(i));
+            if (ids.get(i) == insertedId) {
+                sendUpdate(curValue, RowKind.INSERT, out);
+            } else {
+                sendUpdateBefore(curValue, prevAcc, out);
+                sendUpdateAfter(curValue, out);
+                if (i != 0 && insertedId == -1) {
+                    iterator.next();
+                }
+            }
+            i++;
+        }
+    }
+
+    private void removeFromSortedList(RowData input, Collector<RowData> out) throws Exception {
+        boolean isRetracted = false;
+        input.setRowKind(RowKind.INSERT);
+        List<RowData> sortedKeyList = sortedKeyState.value();
+        ListIterator<RowData> iterator = sortedKeyList.listIterator();
+
+        while (iterator.hasNext()) {
+            RowData curKey = iterator.next();
+            RowData curValue = valueMapState.get(curKey.getLong(keyIdx));
+            if (isRetracted) {
+                List<Long> ids = sortIndexState.get(curKey.getLong(1));
+                if (ids.size() > 1) {
+                    updateSameSortKeys(iterator, ids, -1L, out);
+                } else {
+                    // Update single row
+                    if (!updateRow(curKey, out)) {
+                        break;
+                    }
+                }
+            } else if (equaliser.equals(curValue, input)) {
+                List<Long> ids = sortIndexState.get(curKey.getLong(1));
+                if (ids.size() > 1) {
+                    removeRowFromSameSortKeys(sortedKeyList, curKey, iterator, ids, out);
+                } else {
+                    // Retract single record
+                    RowData curAcc = accMapState.get(curKey.getLong(keyIdx));
+                    sendUpdateBefore(input, curAcc, out);
+                    iterator.remove();
+                    removeFromState(sortedKeyList, ids, curKey.getLong(keyIdx), -1, curKey.getLong(1));
+                    if (iterator.hasPrevious()) {
+                        setAccumulatorOfPrevRow(sortedKeyList, iterator.previousIndex());
+                    } else {
+                        currFunction.setAccumulators(currFunction.createAccumulators());
+                    }
+                }
+                isRetracted = true;
+            }
+        }
+    }
+
+    private void removeRowFromSameSortKeys(List<RowData> sortedKeyList, RowData curKey, ListIterator<RowData> iterator, List<Long> ids, Collector<RowData> out)
+            throws Exception {
+        int i = 0;
+        int removeIndex = -1;
+        while (i < ids.size()) {
+            if (ids.get(i) == curKey.getLong(keyIdx)) {
+                removeIndex = i;
+                break;
+            }
+            i++;
+        }
+
+        setAccumulatorOfPrevRow(sortedKeyList, iterator.nextIndex() - 1 - i  - 1);
```

Comment:
```suggestion
        setAccumulatorOfPrevRow(sortedKeyList, iterator.nextIndex() - 1 - removeIndex  - 1);
```
This is quite cryptic. A comment would help

Reply:
Removed this and no longer exists in the new implementation.